csrc/abstract_tensor_schedule.cpp

```diff
@@ -0,0 +1,460 @@
+    //     ValGroup 3: iS3
+    //     ValGroup 4: iS4
+    //     ValGroup 5: iS5
+    //     ValGroup 6: iS7
+    //     ValGroup 7: iS8
+    //     ExprGroup 0:
+    //       iS3 = merge(iS0, iS1)
+    //     ExprGroup 1:
+    //       iS4, iS5 = split(iS3, 16)
+    //     ExprGroup 2:
+    //       iS8 = merge(iS6, iS7)
+    //
+    //   abstract_tensor_.domain: ValGroup 4, ValGroup 5, ValGroup 7
+    //
+    //      VG0   VG1
+    //        \   /
+    //         EG0
+    //          |
+    //         VG3       VG2   VG6
+    //          |          \   /
+    //         EG1          EG2
+    //        /   \          |
+    //      VG4   VG5       VG7
+    //
+    // In this case, tv has loop domains in ValGroups 0, 1, and 2. IterDomains
+    // iS6, iS7, and iS8 might be associated to another tensor in the fusion.
+    //
+    // ValGroups 4 and 5 are computable since those ValGroups are produced by
+    // ExprGroup 1 which itself produced by ExprGroup 0, and tv includes iS0 and
+    // iS1.
+    //
+    // However, ValGroup 7 is not computable. It is produced by ExprGroup 2
+    // whose input ValGroups are 2 and 6. ValGroup 2 is computable since iS2 is
+    // in tv, however there is no IterDomain in tv that can be used to represent
+    // ValGroup 6 which also has no producer ValGroups.
+
+    // Any computable ValGroup should have an entry in computed_ids. If we prove
+    // that the ValGroup is not computable, we insert it here.
+    std::unordered_set<ValGroup> uncomputable_groups;
+
+    // We are trying to evaluate the ValGroup g which acts like a symbolic
+    // IterDomain using the concrete IterDomains found in the TensorView. To do
+    // so, we recursively evaluate the input ValGroups then create a new Expr
+    // for the corresponding ExprGroup. This recursion is implemented using a
+    // stack of ValGroups that are left to be evaluated.
+    std::stack<ValGroup> eval_stack({g});
+    while (!eval_stack.empty()) {
+      ValGroup vg = eval_stack.top();
+      eval_stack.pop();
+      if (computed_ids.count(vg) != 0) {
+        // vg is already computed
+        continue;
+      }
+
+      // We need to try and produce an ID for the group vg, so we look at
+      // definitions for the Vals in vg and check whether we have their inputs
+      // computed yet (computed_ids) and whether they can be computed
+      // (uncomputable_groups). If an ExprGroup does not have any computable
+      // inputs, we move on to the next ExprGroup. There might be inputs that
+      // have not yet been computed or proven uncomputable; in those cases we
+      // will place them into the vector below so that we can push them onto
+      // the stack so that we can try again.
+      std::vector<ValGroup> unprocessed_producer_groups;
+      // This will be set to true once we have computed an IterDomain for this
+      // ValGroup
+      bool vg_computed = false;
+      for (const ExprGroup& eg : graph_->getDefinitions(vg)) {
```

Comment:
IIUC, checking multiple definitions would matter because there could be multiple different exprs in some graphs like the Permissive graph. If this graph is the Exact graph, there should be only one ExprGroup, so this loop should not matter. Am I understanding correctly?